gax-java/gax/src/main/java/com/google/api/gax/rpc/RequestParamsBuilder.java

```diff
@@ -56,19 +59,49 @@ public static RequestParamsBuilder create() {
    * pre-configured path templates. This method is called repeatedly for each configured routing
    * rule parameter, it's possible that the incoming field value from request is null or there is no
    * matches found, we'll continue the match-and-extract process for the next routing rule parameter
-   * in such case.
+   * in such case. This method will percent encode both the header key and header value.
    *
    * @param fieldValue the field value from a request
    * @param headerKey the header key for the routing header param
    * @param pathTemplate {@link PathTemplate} the path template used for match-and-extract
    */
   public void add(String fieldValue, String headerKey, PathTemplate pathTemplate) {
-    if (fieldValue == null) {
+    if (fieldValue == null || fieldValue.isEmpty()) {
       return;
     }
     Map<String, String> matchedValues = pathTemplate.match(fieldValue);
     if (matchedValues != null && matchedValues.containsKey(headerKey)) {
-      paramsBuilder.put(headerKey, matchedValues.get(headerKey));
+      String encodedKey = percentEncodeString(headerKey);
+      String encodedValue = percentEncodeString(matchedValues.get(headerKey));
+      paramsBuilder.put(encodedKey, encodedValue);
+    }
+  }
+
+  /**
+   * Add an entry to paramsBuilder with key-value pairing of (headerKey, fieldValue). The only
+   * validation done is to ensure the headerKey and fieldValue are not null and non-empty. This
+   * method will percent encode both the header key and header value.
+   *
+   * @param headerKey the header key for the routing header param
+   * @param fieldValue the field value from a request
+   */
+  public void add(String headerKey, String fieldValue) {
+    if (headerKey == null || headerKey.isEmpty() || fieldValue == null || fieldValue.isEmpty()) {
+      return;
+    }
+    String encodedKey = percentEncodeString(headerKey);
+    String encodedValue = percentEncodeString(fieldValue);
+    paramsBuilder.put(encodedKey, encodedValue);
+  }
+
+  // Percent encode the value passed in
+  private String percentEncodeString(String value) {
+    try {
+      return URLEncoder.encode(value, StandardCharsets.UTF_8.toString());
```

Comment:
Noticed that there are some [recommendations](https://stackoverflow.com/questions/29276432/what-character-set-should-be-used-for-url-encoding) against URLEncoder, can you take a look?

Reply:
I see. I think we could use Guava's [PercentEncoder](https://github.com/google/guava/blob/master/guava/src/com/google/common/net/PercentEscaper.java). This should properly encode space characters and other chars (i.e. `+` and `/` and `.`)

Reply:
On a second look, it seems that I'll need to percent-encode the reserved characters: https://en.wikipedia.org/wiki/URL_encoding

The PercentEscaper seems to encode every chars so I'll whitelist these tokens `.-~_` using:
`new PercentEscaper(".-~_" , false)` and I believe it should work.

Reply:
> Unable to create tables with the unreserved characters and ~ and .

Is it because the backend does not allow these characters to be used for creating tables or the encoding? I guess the former but want to make sure.

Reply:
Yep, the former. This is the bigquery restriction that I see: `Unicode letters, marks, numbers, connectors, dashes or spaces allowed.`